pkg/server/image_pull.go

```diff
@@ -17,14 +17,226 @@ limitations under the License.
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("readblob failed for digest:%q err: %v", image.Target.Digest, err)
+	}
+	err = json.Unmarshal(p, &manifest)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("unmarshal blob to manifest failed for digest:%q %v", image.Target.Digest, err)
+	}
+	size, err = image.Size(ctx, c.contentProvider)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("size failed for image:%q %v", image.Target.Digest, err)
+	}
+	compressedSize = uint64(size)
+	return resolvedImageName, manifest, compressedSize, nil
+}
+
+func (c *criContainerdService) pullImage(ctx context.Context, ref string) (imagespec.Descriptor, int64, error) {
+	var (
+		size              int64
+		desc              imagespec.Descriptor
+		resolvedImageName string
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler for a sequence of handlers which: 1) fetch the object using a
+	// FetchHandler; and 3) recurse through any sub-layers via a ChildrenHandler
+	resolver := docker.NewResolver()
+	eg, egctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		var (
+			egerr   error
+			fetcher remotes.Fetcher
+		)
+
+		resolvedImageName, desc, fetcher, egerr = resolver.Resolve(egctx, ref)
+		if egerr != nil {
+			return egerr
+		}
+
+		eg.Go(func() error {
+			return c.imageStore.Put(egctx, resolvedImageName, desc)
+		})
+
+		return containerd_images.Dispatch(
+			egctx,
+			containerd_images.Handlers(
+				remotes.FetchHandler(c.contentIngester, fetcher),
+				containerd_images.ChildrenHandler(c.contentProvider)),
+			desc)
+	})
+
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		errs <- eg.Wait()
+	}()
+
+	select {
+	case err := <-errs:
+		if err != nil {
+			return desc, size, err
+		}
+	case <-egctx.Done():
+	}
+
+	image, err := c.imageStore.Get(ctx, resolvedImageName)
+	if err != nil {
+		return desc, size,
+			fmt.Errorf("get failed for image:%q err: %v", resolvedImageName, err)
+	}
+	p, err := content.ReadBlob(ctx, c.contentProvider, image.Target.Digest)
+	if err != nil {
+		return desc, size,
+			fmt.Errorf("readblob failed for digest:%q err: %v", image.Target.Digest, err)
+	}
+	var manifest imagespec.Manifest
+	err = json.Unmarshal(p, &manifest)
+	if err != nil {
+		return desc, size,
+			fmt.Errorf("unmarshal blob to manifest failed for digest:%q %v", image.Target.Digest, err)
+	}
+	_, err = c.rootfsUnpacker.Unpack(ctx, manifest.Layers) // ignoring returned chainID for now
+	if err != nil {
+		return desc, size,
+			fmt.Errorf("unpack failed for manifest layers:%v %v", manifest.Layers, err)
+	}
+	_, err = image.RootFS(ctx, c.contentProvider) // ignoring diffIDs for now TODO(mikebrow): probably shouldn't even call it if not using them
```

Comment:
IIUC, this function only calculate the chainID. If you don't need the chainID, then just remove this function.